File: golang/data_structures/linked_list/linked_list.go
package linked_list

import (
	"fmt"
	"strconv"
	"strings"
)

type Node struct {
	value int
	next  *Node
}

func NewNode(value int) *Node {
	node := new(Node)
	node.value = value

	return node
}

type LinkedList struct {
	head *Node
	tail *Node
}

func NewLinkedList(args ...int) *LinkedList {
	list := new(LinkedList)

	return list
}

func (l *LinkedList) append(value int) {
	if l.head == nil {
		l.head = NewNode(value)
		l.tail = l.head
	} else {
		l.tail.next = NewNode(value)
		l.tail = l.tail.next
	}
}

func (l *LinkedList) search(value int) (foundValue int, ok bool) {
	node := l.head

	if node == nil {
		return 0, false
	}

	for node.next != nil {
		if node.value == value {
			return node.value, true
		}
		node = node.next
	}
	return 0, false
}

func (l *LinkedList) index(value int) (index int, ok bool) {
	node := l.head
	index = 0

	if node == nil {
		return index, false
	}

	for node.next != nil {
		if node.value == value {
			return index, true
		}
		node = node.next
		index++
	}
	return index, false
}

func (l *LinkedList) String() string {
	node := l.head

	var values []string

	if node == nil {
		return "[]"
	}

	for node.next != nil {
		values = append(values, strconv.Itoa(node.value))

		node = node.next
	}
	values = append(values, strconv.Itoa(node.value))

	return fmt.Sprintf("[%s]", strings.Join(values, ", "))
}

func (l *LinkedList) Len() int {
	node := l.head

	var length int

	if node == nil {
		return 0
	}

	for node.next != nil {
		length++

		node = node.next
	}
	length++
	return length
}

File: data_structures/linked_list.py
from data_structures.node import Node


class LinkedList:
    head = None
    tail = None

    def __init__(self, lst=None):
        lst = lst or []

        for value in lst:
            self.append(value)

    def append(self, value):
        if self.head is None:
            self.head = Node(value)
            self.tail = self.head
        else:
            self.tail.next = Node(value)
            self.tail = self.tail.next

    def search(self, value):
        node = self.head

        while node.next is not None:
            if node.data == value:
                return node

            node = node.next

    def insert_after(self, node, value):
        n = self.head

        while n.next is not None:
            if n == node:
                new_node = Node(value)
                new_node.next = n.next
                n.next = new_node
                return

            n = n.next

    def insert(self, pos, value):
        node = self.head
        index = 0

        while node:
            if index == pos - 1:
                new_node = Node(value)
                new_node.next = node.next
                node.next = new_node
                return True

            node = node.next
            index += 1
        return False

    def index(self, value):
        node = self.head
        index = 0

        while node:
            if node.data == value:
                return index

            node = node.next
            index += 1
        return None

    def __bool__(self):
        return self.head is not None

    def __iter__(self):
        node = self.head

        if node is None:
            return

        while node.next is not None:
            yield node.data

            node = node.next

        yield node.data

    def __len__(self):
        counter = 0
        for _ in self:
            counter += 1

        return counter

    def __repr__(self):
        return str([x for x in self])
